PWGJE/EMCALJetTasks/UserTasks/AliAnalysisTaskEmcalVsPhos.cxx

```diff
@@ -1041,6 +1169,145 @@ void AliAnalysisTaskEmcalVsPhos::FillClusterHistograms()
+
+  }
+  
+  // Loop through jets, to fill various histograms
+  AliJetContainer* jets = GetJetContainer(0); // there is only a single, det-level jet finder here
+  for (const auto jet : jets->accepted()) {
+    
+    Double_t jetPt = GetJetPt(jet, 0);
+    
+    // Keep track of hadronic calo energy in each jet
+    Double_t hadCaloEnergyUnmatched = 0;
+    Double_t hadCaloEnergyMatchedNonlincorr = 0;
+    Double_t hadCaloEnergyMatchedHadCorr = 0;
+    
+    // Loop through clusters in each jet, to plot several histograms
+    Int_t nClusters = jet->GetNumberOfClusters();
+    for (Int_t iClus = 0; iClus < nClusters; iClus++) {
+      
+      clus = jet->Cluster(iClus);
+      
+      // Get the particle type of the cluster
+      ParticleType particleType1 = kUndefined;
+      Int_t label = TMath::Abs(clus->GetLabel());
+      if (label > 0) {
+        particleType1 = GetParticleType1(clus, mcevent, clusters->GetArray());
+      }
+      
+      // Plot M02 for each particle type
+      histname = "JetPerformance/hM02VsParticleTypeJets";
+      Double_t x[4] = {clus->GetM02(), clus->GetNonLinCorrEnergy(), particleType1, jetPt};
+      fHistManager.FillTHnSparse(histname, x);
+      
+      // If the cluster is a hadron, sum its energy to compute the jet's hadronic calo energy
+      if (particleType1 == kHadron) {
+        Bool_t hasMatchedTrack = (clus->GetNTracksMatched() > 0);
+        if (hasMatchedTrack) {
+          hadCaloEnergyMatchedNonlincorr += clus->GetNonLinCorrEnergy();
+          hadCaloEnergyMatchedHadCorr += clus->GetHadCorrEnergy();
+        }
+        else {
+          hadCaloEnergyUnmatched += clus->GetNonLinCorrEnergy();
+        }
+      }
+
+    }
+    
+    // Fill hadronic calo energy in each jet
+    
+    // (Jet pT, Summed energy of hadronic clusters without a matched track)
+    histname = "JetPerformance/hHadCaloEnergyUnmatched";
+    fHistManager.FillTH2(histname, jetPt, hadCaloEnergyUnmatched);
+    
+    // (Jet pT vs. Summed energy of hadronic clusters with a matched track (before hadronic correction))
+    histname = "JetPerformance/hHadCaloEnergyMatchedNonlincorr";
+    fHistManager.FillTH2(histname, jetPt, hadCaloEnergyMatchedNonlincorr);
+    
+    // (Jet pT vs. Summed energy of hadronic clusters with a matched track (after hadronic correction))
+    histname = "JetPerformance/hHadCaloEnergyMatchedHadCorr";
+    fHistManager.FillTH2(histname, jetPt, hadCaloEnergyMatchedHadCorr);
+    
+    // Loop through particle types, and plot jet composition for each particle type
+    histname = "JetPerformance/hJetComposition";
+    for (Int_t type = 0; type < 8; type++) {
+      
+      ParticleType particleType1 = kUndefined;
+      Double_t nSum = 0;
+      Double_t pTsum = 0;
+      
+      // Loop through clusters in jet, and add to sum if cluster matches particle type
+      for (Int_t iClus = 0; iClus < nClusters; iClus++) {
```

Comment:
same as above : `for(auto cluster : clusCont->accepted())`

Reply:
This will be fixed with the constituent framework to be pushed soon. Then you can iterate over all clusters / tracks in a jet.